runtime/lua/vim/lsp/util.lua

```diff
@@ -1304,27 +1324,72 @@ function M.open_floating_preview(contents, syntax, opts)
     opts = { opts, 't', true };
   }
   opts = opts or {}
+  opts.stylize_markdown = opts.stylize_markdown == nil and true or opts.stylize_markdown
+  opts.close_events = opts.close_events or {"CursorMoved", "CursorMovedI", "BufHidden", "InsertCharPre"}
+
+  local bufnr = api.nvim_get_current_buf()
+
+  -- check if this popup is focusable and we need to focus
+  if opts.focus_id then
+    -- Go back to previous window if we are in a focusable one
+    local current_winnr = api.nvim_get_current_win()
+    if npcall(api.nvim_win_get_var, current_winnr, opts.focus_id) then
+      api.nvim_command("wincmd p")
+      return bufnr, current_winnr
+    end
+    do
+      local win = find_window_by_var(opts.focus_id, bufnr)
+      if win and api.nvim_win_is_valid(win) and vim.fn.pumvisible() == 0 then
+        -- focus and return the existing buf, win
+        api.nvim_set_current_win(win)
+        api.nvim_command("stopinsert")
+        return api.nvim_win_get_buf(win), win
+      end
+    end
+  end
+
+  local floating_bufnr = api.nvim_create_buf(false, true)
+  local do_stylize = syntax == "markdown" and opts.stylize_markdown
+
 
   -- Clean up input: trim empty lines from the end, pad
   contents = M._trim(contents, opts)
 
+  if do_stylize then
+    contents = M.stylize_markdown(floating_bufnr, contents, opts)
+  elseif syntax then
+    api.nvim_buf_set_option(floating_bufnr, 'syntax', syntax)
+    api.nvim_buf_set_lines(floating_bufnr, 0, -1, true, contents)
+  end
+
   -- Compute size of float needed to show (wrapped) lines
   opts.wrap_at = opts.wrap_at or (vim.wo["wrap"] and api.nvim_win_get_width(0))
   local width, height = M._make_floating_popup_size(contents, opts)
 
-  local floating_bufnr = api.nvim_create_buf(false, true)
-  if syntax then
-    api.nvim_buf_set_option(floating_bufnr, 'syntax', syntax)
-  end
   local float_option = M.make_floating_popup_options(width, height, opts)
   local floating_winnr = api.nvim_open_win(floating_bufnr, false, float_option)
-  if syntax == 'markdown' then
+  if do_stylize then
     api.nvim_win_set_option(floating_winnr, 'conceallevel', 2)
+    api.nvim_win_set_option(floating_winnr, 'concealcursor', 'n')
   end
-  api.nvim_buf_set_lines(floating_bufnr, 0, -1, true, contents)
+  -- disable folding
+  api.nvim_win_set_option(floating_winnr, 'foldenable', false)
+  -- soft wrapping
+  api.nvim_win_set_option(floating_winnr, 'wrap', true)
+  api.nvim_win_set_option(floating_winnr, 'linebreak', true)
+  api.nvim_win_set_option(floating_winnr, 'breakindent', true)
+  api.nvim_win_set_option(floating_winnr, 'breakindentopt', "shift:2")
```

Comment:
@mjlbach see the folding and wrapping options above. If any other options would make sense, let me kow.

Reply:
I think this is good? Seems to address the current open issues/complaints, we can also make more iterative changes if people give feedback.

Reply:
yep, agree 👍

Reply:
Let me test those changes; they seem to be a bit intrusive for me.
(I use `floating_preview` to preview a location in a file, and I would like to see that as close to what my normal buffer is showing me. Previously, this was done by a logic that takes `wrap` from the buffer itself while still allowing it to be overriden.)

Reply:
@clason maybe I should add an option for the wrapping?

Reply:
there's wrapping code in there already. I'll look into that first.

Reply:
previously we had
```lua
  opts.wrap_at = opts.wrap_at or (vim.wo["wrap"] and api.nvim_win_get_width(0))
```
which is then passed to a `_make_floating_popup_size` function. (Took a few iterations to get it working right, too ;))

Reply:
@clason can you clarify the regression? I don't see a case in which we would want a floating window to not have wrapped text.

Reply:
as I said, I haven't tested it yet but want to to make sure (not at my dev machine atm). What struck me was not the fact that there is wrapping (although some people might not like it; I'll look at previous issues on that) but the options for styling (`breakindent`, with specific width).

Reply:
That `opts.wrap_at` code looks really odd. It also never sets actually wrapping on the floating window. It's only used to calculate the height of the popup based of the `wrap` property of the parent window. (which is probably inherited by the float?)

I'll change it so that the `opts.wrap_at` defaults to true.

Reply:
There was https://github.com/neovim/neovim/issues/12998, for example

Reply:
@clason yeah, I wasn't sure about the `breakindent`. Maybe I should leave that out for now?

Reply:
It's only used to calculate the height of the popup based of the `wrap` property of the parent window. (which is probably inherited by the float?)

Yes.

Reply:
Give me five minutes to check out the current version.

Reply:
Yeah, I think that's not going to be universally popular. I'd suggest not setting these options (`linebreak` and `breakindent`) so that they are -- if set -- inherited by the user's config. That seems to work well for `hover` and `preview_location` for me.

Reply:
(And I'm even using almost the same options apart from the `breakindentopt`, and it's still bugging me ;))

Reply:
@clason I just pushed some changes that probably make more sense with regards to `wrap`:

* `open_floating_preview` has a new `wrap` option that defaults to true
* if `wrap=true`, then we use `wrap_to` to calculate the height and width **AND** we enable wrapping on the float

Reply:
That still sets unwanted options (which were my main point of contention, not the wrapping itself) ;)

Reply:
Philosophically I think it's ok to be opinionated on these windows and not add too many options (which increases complexity) as a lot of people are just going to copy/past and override the handler anyways.

Reply:
To be clear: even _if_ I want wrapping in my floats (which I do), I don't want _your_ options; the float should respect the options I set in my `init.vim` (or not). Which seems to work nicely if the options are simply not set here.

Reply:
@clason ok, I'll remove the indent stuff. 
But maybe keep the `linebreak`? (wrap around words)

Reply:
@clason I get your point. I'll remove the indent and linebreak, but will keep the `wrap`

Reply:
@clason done

Reply:
Why? If people want linebreak, they can set it in their `init.vim`. If they don't, they can't easily override it.

(I disagree with the "opinionated windows" since this code is not just used to show simple `hover` text but also complete previews of parts of my buffer.)

Reply:
@mjlbach To be clear, I'm not saying "add options and overrides", I'm saying "don't set options and let `init.vim` handle this".

Reply:
Sure, I haven't actually used code preview beyond testing it, so I'm not sure what the implications are of setting options that may/should be inherited from global neovim settings.

Reply:
I agree. So now it will still enable `wrap` on the window, but won't set `linebreak` or `breakindent`. Those setting will now simply be based off the options the user has set in their init.{vim,lua}

Reply:
(Sorry, was typing while you commented; didn't want to beat a dead horse. Now it's fine, thank you!)